cloudify_aws/ec2/keypair.py

```diff
@@ -0,0 +1,263 @@
+
+# Cloudify imports
+from cloudify_aws import utils, constants
+from cloudify import ctx
+from cloudify.exceptions import NonRecoverableError
+from cloudify.decorators import operation
+from cloudify_aws.base import AwsBaseNode
+
+
+@operation
+def creation_validation(**_):
+    return KeyPair().creation_validation()
+
+
+@operation
+def create(args=None, **_):
+    return KeyPair().created(args)
+
+
+@operation
+def delete(args=None, **_):
+    return KeyPair().deleted(args)
+
+
+class KeyPair(AwsBaseNode):
+
+    def __init__(self):
+        super(KeyPair, self).__init__(
+                constants.KEYPAIR['AWS_RESOURCE_TYPE'],
+                constants.KEYPAIR['REQUIRED_PROPERTIES']
+        )
+        self.not_found_error = constants.KEYPAIR['NOT_FOUND_ERROR']
+        self.get_all_handler = {
+            'function': self.client.get_all_key_pairs,
+            'argument': 'keynames'
+        }
+
+    def creation_validation(self, **_):
+        """ This validates all nodes before bootstrap.
+        """
+
+        for property_key in self.required_properties:
+            utils.validate_node_property(property_key, ctx.node.properties)
+
+        key_file = self._get_path_to_key_file()
+        key_file_in_filesystem = self._search_for_key_file(key_file)
+
+        if ctx.node.properties['use_external_resource']:
+            if not key_file_in_filesystem:
+                raise NonRecoverableError(
+                        'External resource, but the key file '
+                        'does not exist locally.')
+            try:
+                self._get_key_pair_by_id(ctx.node.properties['resource_id'])
+            except NonRecoverableError as e:
+                raise NonRecoverableError(
+                        'External resource, '
+                        'but the key pair does not exist in the account: '
+                        '{0}'.format(str(e)))
+        else:
+            if key_file_in_filesystem:
+                raise NonRecoverableError(
+                        'Not external resource, '
+                        'but the key file exists locally.')
+            try:
+                self._get_key_pair_by_id(ctx.node.properties['resource_id'])
+            except NonRecoverableError:
+                pass
+            else:
+                raise NonRecoverableError(
+                        'Not external resource, '
+                        'but the key pair exists in the account.')
+
+        return True
+
+    def create(self, args=None, **_):
+        """Creates a keypair."""
+
+        key_pair_name = utils.get_resource_id()
+
+        kp = self.execute(self.client.create_key_pair,
+                          dict(key_name=key_pair_name), raise_on_falsy=True)
+
+        self._save_key_pair(kp)
+
+        ctx.instance.runtime_properties[constants.AWS_TYPE_PROPERTY] = \
+            constants.KEYPAIR['AWS_RESOURCE_TYPE']
+
+        self.resource_id = kp.name
+
+        return True
+
+    def created(self, args=None):
```

Comment:
Shouldn't need to override this...you override self.use_external_resource_naively